### ISSUE TITLE
pyright complains overload will never be used because parameters overlap

### ISSUE THREAD

**Issue by notatallshaw** [microsoft/pyright#4622]
**Describe the bug**

I'm trying to express two possible scenarios that my function could take as one of the parameters:

1. A type with implements `__radd__` and returns it's own type
2. A type with implements `__radd__` and does not returns it's own type

**To Reproduce**

1. Write a Python file like this:

```python
from typing import Protocol, Any, TypeVar, overload

class RAddable(Protocol):
    def __radd__(self: "RAddable", other: Any, /) -> Any:
        ...

T = TypeVar("T")
class SameRAddable(Protocol):
    def __radd__(self: T, other: Any, /) -> T:
        ...

RA = TypeVar("RA", bound=SameRAddable)

@overload
def add(a: Any, b: RA) -> RA:
    ...

@overload
def add(a: Any, b: RAddable) -> Any:
    ...

def add(a: Any, b: Any) -> Any:
    return a + b
```

2. Run pyright on this file

3. Get the following error on the second overload:

```
Overload 2 for "add" will never be used because its parameters overlap overload 1Pylance[reportOverlappingOverload](https://github.com/microsoft/pyright/blob/main/docs/configuration.md#reportOverlappingOverload)
```

**Expected behavior**

While the second overload does overlap with the first overload, it isn't completely overlapped. In particular if `__radd__` explictly returns a different type, e.g.:

```python
class Foo:
    pass

class Bar:
    def __radd__(self: Self, other: Any) -> str:
        return "Hello World"

print(add(Foo(), Bar()))
```

Bar here clearly does not match overload 1, but does match overload 2.

**Additional context**

It doesn't seem to be possible in current Python type hinting to express the return type of overload 2 as "Not RA", but I do appreciate this might be a design choice and not a bug in pyright.

**Comment by erictraut** [comment]
Thanks for the bug report. This will be addressed in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.295, which I just published. It will also be included in a future release of pylance.
